<!-- ... -->

<launch>
   
  <!-- Choose visualization -->
  <arg name="rviz" default="true" />
  <arg name="rtabmapviz" default="false" />

  <!-- Localization-only mode -->
  <arg name="localization"            default="false"/>

  <!-- Run the ROS package stereo_image_proc for image rectification -->
    <include file="$(find ocams)/launch/ocams_ros.launch" />

  <group ns="/stereo" >
    <node pkg="nodelet" type="nodelet" name="stereo_nodelet"  args="manager"/>
    <node pkg="stereo_image_proc" type="stereo_image_proc" name="stereo_image_proc">
         <param name="disparity_range" value="128"/>
    </node>
  </group>
            
  <group ns="rtabmap">
   
    <!-- Stereo Odometry -->   
    <node pkg="rtabmap_ros" type="stereo_odometry" name="stereo_odometry" output="screen">
      <remap from="left/image_rect"     to="/stereo/left/image_rect"/>
      <remap from="right/image_rect"    to="/stereo/right/image_rect"/>
      <remap from="left/camera_info"    to="/stereo/left/camera_info"/>
      <remap from="right/camera_info"   to="/stereo/right/camera_info"/>
      <remap from="odom"                to="/stereo_odometry"/>

      <param name="Odom/Strategy"      type="string" value="0"/> <!-- 0=Frame-to-Map, 1=Frame=to=Frame -->
      <param name="Odom/FillInfoData"  type="string" value="$(arg rtabmapviz)"/> 
      <param name="OdomF2M/MaxSize"    type="string" value="2000"/> <!-- maximum features map size, default 2000 -->
      <param name="Vis/EstimationType" type="string" value="0"/> <!-- 0=3D->3D 1=3D->2D (PnP) -->
      <param name="Vis/MaxDepth"       type="string" value="20"/>
      <param name="Vis/MinInliers"     type="string" value="10"/>
      <param name="Vis/MaxFeatures"    type="string" value="1000"/> <!-- maximum features extracted by image, default 1000 -->
      <param name="GFTT/MinDistance"   type="string" value="0"/>
      <param name="GFTT/QualityLevel"  type="string" value="0.00001"/>

      <param name="frame_id"                    type="string" value="base_link"/>
      <param name="odom_frame_id"               type="string" value="odom"/>
      <param name="approx_sync"                 type="bool"   value="false"/>

    </node>

    <!-- RTAB-Map --> 
    <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="screen" args="--delete_db_on_start">
      <remap from="left/image_rect"    to="/stereo/left/image_rect_color"/>
      <remap from="right/image_rect"   to="/stereo/right/image_rect"/>
      <remap from="left/camera_info"   to="/stereo/left/camera_info"/>
      <remap from="right/camera_info"  to="/stereo/right/camera_info"/>
      <remap from="odom"               to="/stereo_odometry"/>

      <param name="frame_id"           type="string" value="base_link"/>
      <param name="subscribe_stereo"   type="bool" value="true"/>
      <param name="subscribe_depth"    type="bool" value="false"/>
      <param name="queue_size"         type="int" value="10"/>

      <!-- RTAB-Map's parameters -->
      <param name="Rtabmap/TimeThr"                   type="string" value="700"/>
      <param name="Rtabmap/DetectionRate"             type="string" value="0.0"/>
      <param name="Grid/DepthDecimation"              type="string" value="4"/>
      <param name="Grid/FlatObstacleDetected"         type="string" value="true"/>
      <param name="Kp/MaxFeatures"                    type="string" value="200"/>
      <param name="Kp/MaxDepth"                       type="string" value="10"/>
      <param name="Kp/DetectorStrategy"               type="string" value="0"/>   <!-- use SURF -->
      <param name="SURF/HessianThreshold"             type="string" value="1000"/>
      <param name="Vis/EstimationType"                type="string" value="0"/>   <!-- 0=3D->3D, 1=3D->2D (PnP) -->
      <param name="RGBD/LoopClosureReextractFeatures" type="string" value="true"/>
      <param name="Vis/MaxDepth"                      type="string" value="20"/>

      <!-- localization mode -->
      <param     if="$(arg localization)"    name="Mem/IncrementalMemory" type="string" value="false"/>
      <param unless="$(arg localization)"    name="Mem/IncrementalMemory" type="string" value="true"/>
      <param  name="Mem/InitWMWithAllNodes" type="string" value="$(arg localization)"/>

    </node>

    <!-- Visualisation RTAB-Map -->
    <node if="$(arg rtabmapviz)" pkg="rtabmap_ros" type="rtabmapviz" name="rtabmapviz" args="-d $(find rtabmap_ros)/launch/config/rgbd_gui.ini" output="screen">
      <param name="subscribe_stereo"    type="bool" value="true"/>
      <param name="subscribe_odom_info" type="bool" value="true"/>
      <param name="queue_size"          type="int" value="10"/>
      <param name="frame_id"            type="string" value="base_link"/>
      <param name="approx_sync"         type="bool"   value="false"/>

      <remap from="left/image_rect"     to="/stereo/left/image_rect_color"/>
      <remap from="right/image_rect"    to="/stereo/right/image_rect"/>
      <remap from="left/camera_info"    to="/stereo/left/camera_info"/>
      <remap from="right/camera_info"   to="/stereo/right/camera_info"/>
      <remap from="odom_info"           to="odom_info"/>
      <remap from="odom"                to="/stereo_odometry"/>
      <remap from="mapData"             to="mapData"/>
    </node>
  </group>

  <!-- Visualisation RVIZ --> 
  <node if="$(arg rviz)" pkg="rviz" type="rviz" name="rviz" args="-d $(find ocams)/config/rviz/demo_slam.rviz"/>

</launch>
